src/wp-includes/user.php

```diff
@@ -1990,7 +2047,104 @@ function validate_username( $username ) {
 	 * @param bool   $valid    Whether given username is valid.
 	 * @param string $username Username to check.
 	 */
-	return apply_filters( 'validate_username', $valid, $username );
+	$valid = apply_filters( 'validate_username', true, $username );
+	if ( ! $valid ) {
+		$errors->add( 'user_name', __( 'Sorry, that username is invalid.' ) );
+	}
+
+	if ( $errors->has_errors() && $wp_error ) {
+		return $errors;
+	}
+
+	return ! $errors->has_errors();
+}
+
+/**
+ * Check whether the given username is reserved or not.
+ *
+ * In multisite this check if a signup with the username already exist. If the signup has been registered more than two
+ * days ago it'll be deleted and the username considered available again.
+ *
+ * @since n.e.x.t
+ *
+ * @global wpdb $wpdb WordPress database abstraction object.
+ *
+ * @param string $username Username to check.
+ *
+ * @return bool True is the username is available, false otherwise.
+ */
+function is_username_reserved( $username ) {
+	global $wpdb;
+
+	$reserved = false;
+	if ( is_multisite() ) {
+		$signup = $wpdb->get_row( $wpdb->prepare( "SELECT * FROM $wpdb->signups WHERE user_login = %s", $username ) );
+		if ( null !== $signup ) {
+			$registered_at = mysql2date( 'U', $signup->registered );
+			$now           = time();
+			$diff          = $now - $registered_at;
+			// If registered more than two days ago, cancel registration and let this signup go through.
+			if ( $diff > 2 * DAY_IN_SECONDS ) {
+				$wpdb->delete( $wpdb->signups, array( 'user_login' => $username ) );
+			} else {
+				$reserved = true;
+			}
+		}
+	} else {
+		/**
+		 * Filters whether the username is reserved or not.
+		 *
+		 * @since n.e.x.t
+		 *
+		 * @param bool $reserved Whether given username is reserved.
+		 * @param string $username Username to check.
+		 */
+		$reserved = apply_filters( 'is_username_reserved', $reserved, $username );
+	}
+
+	return $reserved;
+}
+
+/**
+ * Check if a user login is valid.
+ *
+ * This function combine various check that are run on the provided user login :
+ * - {@see validate_username}
+ * - {@see username_exists}
+ * - {@see is_username_reserved}
+ *
+ * @since n.e.x.t
+ *
+ * @param string $user_login User login to validate.
+ * @param WP_Error $errors Existing WP_Error object to use. If null a new WP_Error object will be created and returned.
+ *
+ * @return bool|WP_Error True is the login is valid, WP_Error otherwise.
```

Comment:
```suggestion
 * @param string   $user_login User login to validate.
 * @param WP_Error $errors    Existing WP_Error object to use. If null, a new WP_Error object will be created and returned.
 *
 * @return bool|WP_Error True if the login is valid, WP_Error otherwise.
```

Reply:
Preview spacing is wonky here, but the vars and descriptions should line up.